### PR TITLE
msg/async/dpdk:add commands to obtain the NIC status and statistics

### PR DESCRIPTION
**Pull request by fengchunsong**
Commands are added to obtain the network adapter status and statistics
for debugging network adapter packet loss and mbuf insufficiency issues.

Signed-off-by: Chunsong Feng <fengchunsong@huawei.com>
Reviewed-by: luo rixin <luorixin@huawei.com>
Reviewed-by: Han Fengzhe  <hanfengzhe@hisilicon.com>

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by fengchunsong** [comment]
jenkins test api

**Comment by fengchunsong** [comment]
@cyx1231st please help to review,thanks.

**Comment by fengchunsong** [comment]
@tchaikov please help to review,thanks.

**Comment by fengchunsong** [comment]
@neha-ojha please help to review, thanks

**Comment by fengchunsong** [comment]
@tchaikov please help to review,thanks.

File: src/msg/async/dpdk/DPDK.h [hunk 1]
```diff
@@ -769,7 +774,33 @@ class DPDKDevice {
    */
   int init_port_fini();
 
+  void nic_stats_dump(bufferlist &bl);
+  void nic_xstats_dump(bufferlist &bl);
  private:
+  class XstatSocketHook:public AdminSocketHook {
```
Comment (tchaikov) [inline src/msg/async/dpdk/DPDK.h:780]: might want to define this derived class in the .cc file. and keep an instance of `std::unique_ptr<AdminSocketHook>` as a member variable instead.

File: src/msg/async/dpdk/DPDK.h [hunk 2]
```diff
@@ -769,7 +774,33 @@ class DPDKDevice {
...
+	  DPDKDevice *dev;
+	  public:
```
Comment (tchaikov) [inline src/msg/async/dpdk/DPDK.h:782]: wrong indent.

File: src/msg/async/dpdk/DPDK.h [hunk 3]
```diff
@@ -769,7 +774,33 @@ class DPDKDevice {
...
+	    const cmdmap_t & cmdmap,
+	    Formatter *f,
+	    const bufferlist& inbl,
+	    std::function<void(int,const std::string&,bufferlist&)> on_finish) override {
+	    bufferlist bl;
+	    if (prefix == "show_pmd_stats")
+	      dev->nic_stats_dump(bl);
+	    else if (prefix == "show_pmd_xstats")
+	        dev->nic_xstats_dump(bl);
+	    on_finish(0,"",bl);
```
Comment (tchaikov) [inline src/msg/async/dpdk/DPDK.h:801]: please add space after `,`.

File: src/msg/async/dpdk/DPDK.h [hunk 4]
```diff
@@ -769,7 +774,33 @@ class DPDKDevice {
...
+	    ceph_abort("should use async hook");
+	  }
+	  void call_async(
+	    std::string_view prefix,
+	    const cmdmap_t & cmdmap,
+	    Formatter *f,
+	    const bufferlist& inbl,
+	    std::function<void(int,const std::string&,bufferlist&)> on_finish) override {
+	    bufferlist bl;
+	    if (prefix == "show_pmd_stats")
```
Comment (tchaikov) [inline src/msg/async/dpdk/DPDK.h:797]: please add curly bracket around the statement below.

File: src/msg/async/dpdk/DPDK.cc [hunk 5]
```diff
@@ -642,6 +650,119 @@ DPDKQueuePair::DPDKQueuePair(CephContext *c, EventCenter *cen, DPDKDevice* dev,
...
+    return;
+  }
+  if (cnt_xstats != rte_eth_xstats_get_names(_port_idx,xstats_names,cnt_xstats)) {
+    ldout(cct,0) << "Error: Cannot get xstats lookup" << dendl;
+    free(xstats_names);
+    return;
+  }
+
+  //Get stats themselves
+  xstats = (struct rte_eth_xstat *)malloc(sizeof(struct rte_eth_xstat) * cnt_xstats);
```
Comment (tchaikov) [inline src/msg/async/dpdk/DPDK.cc:744]:
```suggestion
  std::vector<rte_eth_xstat> xstats(cnt_xstats);
```

File: src/msg/async/dpdk/DPDK.cc [hunk 6]
```diff
@@ -642,6 +650,119 @@ DPDKQueuePair::DPDKQueuePair(CephContext *c, EventCenter *cen, DPDKDevice* dev,
...
+  if (xstats == nullptr) {
+    ldout(cct,0) << "Cannot allocate memory for xstats" << dendl;
+    free(xstats_names);
+    return;
+  }
+  if (cnt_xstats != rte_eth_xstats_get(_port_idx, xstats, cnt_xstats)) {
```
Comment (tchaikov) [inline src/msg/async/dpdk/DPDK.cc:750]:
```suggestion
  if (cnt_xstats != rte_eth_xstats_get(_port_idx, xstats.data(), xstats.size())) {
```

File: src/msg/async/dpdk/DPDK.cc [hunk 7]
```diff
@@ -642,6 +650,119 @@ DPDKQueuePair::DPDKQueuePair(CephContext *c, EventCenter *cen, DPDKDevice* dev,
...
+    ldout(cct,0)<< "Error: Unable to get xstats" << dendl;
+    free(xstats_names);
+    free(xstats);
+    return;
+  }
+
+  //Display xstats
```
Comment (tchaikov) [inline src/msg/async/dpdk/DPDK.cc:757]: please add space after `//`. but IMHO, we can dispense with this comment. as the code is self-explaining.

File: src/msg/async/dpdk/DPDK.h [hunk 8]
```diff
@@ -769,7 +774,33 @@ class DPDKDevice {
...
  private:
+  class XstatSocketHook:public AdminSocketHook {
+	  DPDKDevice *dev;
+	  public:
+	  explicit XstatSocketHook(DPDKDevice *dev):dev(dev){}
+	  int call(std::string_view prefix,const cmdmap_t& cmdmap,
+                   Formatter *f,
+                   std::ostream& ss,
+                   bufferlist& out) override {
+	    ceph_abort("should use async hook");
```
Comment (tchaikov) [inline src/msg/async/dpdk/DPDK.h:788]: why? i think the synchronized version fits our needs already.

File: src/msg/async/dpdk/DPDK.cc [hunk 9]
```diff
@@ -642,6 +650,119 @@ DPDKQueuePair::DPDKQueuePair(CephContext *c, EventCenter *cen, DPDKDevice* dev,
...
+
+  //Get stats themselves
+  xstats = (struct rte_eth_xstat *)malloc(sizeof(struct rte_eth_xstat) * cnt_xstats);
+  if (xstats == nullptr) {
+    ldout(cct,0) << "Cannot allocate memory for xstats" << dendl;
+    free(xstats_names);
+    return;
+  }
+  if (cnt_xstats != rte_eth_xstats_get(_port_idx, xstats, cnt_xstats)) {
+    ldout(cct,0)<< "Error: Unable to get xstats" << dendl;
```
Comment (tchaikov) [inline src/msg/async/dpdk/DPDK.cc:751]: again, please add a space after `,`. also, i'd suggest use `ldout(cct, 1)` for this error.

File: src/msg/async/dpdk/DPDK.cc [hunk 10]
```diff
@@ -642,6 +650,119 @@ DPDKQueuePair::DPDKQueuePair(CephContext *c, EventCenter *cen, DPDKDevice* dev,
...
+    free(xstats_names);
+    free(xstats);
+    return;
+  }
+
+  //Display xstats
+  for (int i = 0;i < cnt_xstats;i++){
```
Comment (tchaikov) [inline src/msg/async/dpdk/DPDK.cc:758]: add a space after `;`.

File: src/msg/async/dpdk/DPDK.cc [hunk 11]
```diff
@@ -642,6 +650,119 @@ DPDKQueuePair::DPDKQueuePair(CephContext *c, EventCenter *cen, DPDKDevice* dev,
...
+  size_t rx_free_cnt=0;
+
+  for (auto &qp: _queues) {
+    tx_fragments += qp->perf_logger->get(l_dpdk_qp_tx_fragments);
+    rx_fragments += qp->perf_logger->get(l_dpdk_qp_rx_fragments);
+    tx_free_cnt += qp->_tx_buf_factory.ring_size();
+    rx_free_cnt += rte_mempool_avail_count(qp->_pktmbuf_pool_rx);
+  }
+  rte_eth_stats_get(_port_idx,&stats);
+  len = sprintf(buf,"\n %s NIC statistics for port %-2d %s\n",
```
Comment (tchaikov) [inline src/msg/async/dpdk/DPDK.cc:676]: i'd suggest use `formatter` for printing the structured information.

File: src/msg/async/dpdk/DPDK.cc [hunk 12]
```diff
@@ -642,6 +650,119 @@ DPDKQueuePair::DPDKQueuePair(CephContext *c, EventCenter *cen, DPDKDevice* dev,
...
+  static const char *nic_stats_border = "########################"
+  static uint64_t prev_pkts_rx[RTE_MAX_ETHPORTS];
+  static uint64_t prev_pkts_tx[RTE_MAX_ETHPORTS];
+  static uint64_t prev_cycles[RTE_MAX_ETHPORTS];
+  uint64_t diff_pkts_rx, diff_pkts_tx, diff_cycles;
+  uint64_t mpps_rx, mpps_tx;
+  struct rte_eth_stats stats;
+  char buf[LINE_MAX];
+  int len;
+  size_t tx_fragments=0;
```
Comment (tchaikov) [inline src/msg/async/dpdk/DPDK.cc:664]: please add spaces around `=`.

File: src/msg/async/dpdk/DPDK.cc [hunk 13]
```diff
@@ -644,6 +669,90 @@ DPDKQueuePair::DPDKQueuePair(CephContext *c, EventCenter *cen, DPDKDevice* dev,
...
+  f->dump_unsigned("TX-bytes", stats.obytes);
+  f->dump_unsigned("TX-errors", stats.oerrors);
+
+  f->dump_unsigned("RX-nombuf", stats.rx_nombuf);
+  f->dump_unsigned("RX-avail mbufs", rx_free_cnt);
+  f->dump_unsigned("TX-avail mbufs", tx_free_cnt);
+  diff_cycles = prev_cycles[_port_idx];
+  prev_cycles[_port_idx] = rte_rdtsc();
+  if (diff_cycles > 0)
+    diff_cycles = prev_cycles[_port_idx] - diff_cycles;
```
Comment (tchaikov) [inline src/msg/async/dpdk/DPDK.cc:710]: please guard this statement with curly brackets.

File: src/msg/async/dpdk/DPDK.cc [hunk 14]
```diff
@@ -644,6 +669,90 @@ DPDKQueuePair::DPDKQueuePair(CephContext *c, EventCenter *cen, DPDKDevice* dev,
...
+
+  diff_pkts_rx = (stats.ipackets > prev_pkts_rx[_port_idx]) ?
+	         (stats.ipackets - prev_pkts_rx[_port_idx]) : 0;
+  diff_pkts_tx = (stats.opackets > prev_pkts_tx[_port_idx]) ?
+	         (stats.opackets - prev_pkts_tx[_port_idx]) : 0;
+  prev_pkts_rx[_port_idx] = stats.ipackets;
+  prev_pkts_tx[_port_idx] = stats.opackets;
+  mpps_rx = diff_cycles > 0 ? diff_pkts_rx * rte_get_tsc_hz()/diff_cycles : 0;
```
Comment (tchaikov) [inline src/msg/async/dpdk/DPDK.cc:718]: could you please define variable at where it is used for the first time, for better readability?

File: src/msg/async/dpdk/DPDK.cc [hunk 15]
```diff
@@ -644,6 +669,90 @@ DPDKQueuePair::DPDKQueuePair(CephContext *c, EventCenter *cen, DPDKDevice* dev,
...
+
+void DPDKDevice::nic_xstats_dump(Formatter *f)
+{
+  int cnt_xstats;
+
+  // Get count
+  cnt_xstats = rte_eth_xstats_get_names(_port_idx, NULL, 0);
+  if (cnt_xstats < 0) {
+    ldout(cct, 1) << "Error: Cannot get count of xstats" << dendl;
+    return ;
```
Comment (tchaikov) [inline src/msg/async/dpdk/DPDK.cc:733]: please remove the space before `;`

File: src/msg/async/dpdk/DPDK.cc [hunk 16]
```diff
@@ -644,6 +669,90 @@ DPDKQueuePair::DPDKQueuePair(CephContext *c, EventCenter *cen, DPDKDevice* dev,
...
+  rte_eth_stats_get(_port_idx, &stats);
+  f->open_object_section("stats");
+  f->dump_unsigned("RX-packets", stats.ipackets);
+  f->dump_unsigned("DPDK RX-packets", rx_fragments);
+  f->dump_unsigned("RX-bytes", stats.ibytes);
+  f->dump_unsigned("RX-missed", stats.imissed);
+  f->dump_unsigned("RX-errors", stats.ierrors);
+
+  f->dump_unsigned("TX-packets", stats.opackets);
+  f->dump_unsigned("DPDK TX-packets", tx_fragments);
```
Comment (tchaikov) [inline src/msg/async/dpdk/DPDK.cc:700]: i am not sure if it's valid to have a key with space(s) in it. could you use lower-case names and use underscores as delimiters like "rx_bytes" and "dpdk_tx_packets", or have dedicated sections for "rx" and "tx" respectively?

File: src/msg/async/dpdk/DPDK.h [hunk 17]
```diff
@@ -478,6 +479,10 @@ class DPDKQueuePair {
       _ring.push_back(buf);
     }
 
+    int ring_size(){
```
Comment (tchaikov) [inline src/msg/async/dpdk/DPDK.h:482]:
```suggestion
    unsigned ring_size() const {
```

File: src/msg/async/dpdk/DPDK.cc [hunk 18]
```diff
@@ -392,6 +392,23 @@ void DPDKDevice::set_hw_flow_control()
...
+ public:
+  explicit XstatSocketHook(DPDKDevice *dev) : dev(dev) {}
+  int call(std::string_view prefix, const cmdmap_t& cmdmap,
+           Formatter *f,
+           std::ostream& ss,
+           bufferlist& out) override {
+    if (prefix == "show_pmd_stats") {
+      dev->nic_stats_dump(f);
+    } else if (prefix == "show_pmd_xstats") {
+        dev->nic_xstats_dump(f);
```
Comment (tchaikov) [inline src/msg/async/dpdk/DPDK.cc:406]: wrong indent.

File: src/msg/async/dpdk/DPDK.cc [hunk 19]
```diff
@@ -644,6 +669,92 @@ DPDKQueuePair::DPDKQueuePair(CephContext *c, EventCenter *cen, DPDKDevice* dev,
...
+    diff_cycles = prev_cycles[_port_idx] - diff_cycles;
+  }
+
+  uint64_t diff_pkts_rx = (stats.ipackets > prev_pkts_rx[_port_idx]) ?
+	         (stats.ipackets - prev_pkts_rx[_port_idx]) : 0;
+  uint64_t diff_pkts_tx = (stats.opackets > prev_pkts_tx[_port_idx]) ?
+	         (stats.opackets - prev_pkts_tx[_port_idx]) : 0;
+  prev_pkts_rx[_port_idx] = stats.ipackets;
+  prev_pkts_tx[_port_idx] = stats.opackets;
+  uint64_t mpps_rx = diff_cycles > 0 ? diff_pkts_rx * rte_get_tsc_hz()/diff_cycles : 0;
```
Comment (tchaikov) [inline src/msg/async/dpdk/DPDK.cc:722]: please add spaces around `/`

File: src/msg/async/dpdk/DPDK.cc [hunk 20]
```diff
@@ -644,6 +669,92 @@ DPDKQueuePair::DPDKQueuePair(CephContext *c, EventCenter *cen, DPDKDevice* dev,
...
+  std::vector<struct rte_eth_xstat_name> xstats_names(cnt_xstats);
+  if (cnt_xstats != rte_eth_xstats_get_names(_port_idx, xstats_names.data(), cnt_xstats)) {
+    ldout(cct, 1) << "Error: Cannot get xstats lookup" << dendl;
+    return;
+  }
+
+  // Get stats themselves
+  std::vector<struct rte_eth_xstat> xstats(cnt_xstats);
+  if (cnt_xstats != rte_eth_xstats_get(_port_idx, xstats.data(), cnt_xstats)) {
+    ldout(cct, 1)<< "Error: Unable to get xstats" << dendl;
```
Comment (tchaikov) [inline src/msg/async/dpdk/DPDK.cc:748]: pleas add a space before `<<`.

**Review by tchaikov — APPROVED** [review]

**Pull request merged**